blockprod/src/detail/mod.rs

```diff
@@ -141,14 +149,102 @@ impl BlockProduction {
                         })
                     };
 
+                    let block_height = best_block_index.block_height().next_height();
+                    let sealed_epoch_index = chain_config.sealed_epoch_index(&block_height);
+
+                    let sealed_epoch_randomness = sealed_epoch_index
+                        .map(|index| this.get_epoch_data(index))
+                        .transpose()
+                        .map_err(|_| {
+                            ConsensusPoSError::PropertyQueryError(
+                                PropertyQueryError::EpochDataNotFound(block_height),
+                            )
+                        })?
+                        .flatten()
+                        // TODO: no epoch_data means either that no
+                        // epoch has been created yet, or that the
+                        // data is actually missing
+                        .or_else(|| Some(EpochData::new(PoSRandomness::at_genesis(&chain_config))));
+
                     let consensus_data = consensus::generate_consensus_data(
                         &chain_config,
                         &best_block_index,
+                        sealed_epoch_randomness.clone(),
+                        input_data.clone(),
                         block_timestamp,
-                        best_block_index.block_height().next_height(),
+                        block_height,
                         get_ancestor,
-                    );
-                    consensus_data.map(|cons_data| (cons_data, best_block_index))
+                    )?;
+
+                    let finalize_block_data = match input_data {
```

Comment:
I think this match deserves to be a function

Reply:
Yeah, that's one of the places I wanted to clean up later, but while I'm in this code, I might as well add it.

I originally moved it into `/consensus/` but when fixing up the imports I realised that `chainstate` would need to be imported for `ChainstateInterface`, which would up as a circular dependancy... so instead I moved it out of that function to a helper fuction below it.